AX: AXIsolatedObject::setProperty does unsafe access to accessibility-thread only data structure, AXIsolatedTree::m_readerThreadNodeMap, when caching AXProperty::{Font, TextColor}

hoffmanjoshua opened the pull request:
#### c9b5c99e115ef3d76bd4f5240e813e92c4a07047
<pre>
AX: AXIsolatedObject::setProperty does unsafe access to accessibility-thread only data structure, AXIsolatedTree::m_readerThreadNodeMap, when caching AXProperty::{Font, TextColor}
<a href="https://bugs.webkit.org/show_bug.cgi?id=292823">https://bugs.webkit.org/show_bug.cgi?id=292823</a>
<a href="https://rdar.apple.com/151077899">rdar://151077899</a>

Reviewed by Tyler Wilcock.

294634@main made it possible for AXIsoaltedObject::parentObject() to be called on the main thread,
since it started making calls in setProperty in order to determine the parent object&apos;s style.

Instead of using the isoalted object, we can do the same diff-ing of Text Color and Font with live
objects. We need to do this in two places: when updating the property and during initialization.

To ensure we aren&apos;t updating the Font or Text Color property multiple times, we only update once
per subtree that had a Font/Color change, utilizing the parent stack in the style tree resolver.

At the same time, our new way of caching Font and Text Color using ancestors allows us to stop
caching these properties on static text, and instead cache them on any Element. This maintains
our previous optimization, and prevents possible duplication of properties.

This patch was co-authored by Tyler Wilcock.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::setProperty):
(WebCore::defaultTextColor):
(WebCore::getColor):
(WebCore::AXIsolatedObject::cachedTextColor const):
(WebCore::getFont):
(WebCore::AXIsolatedObject::cachedFont const):
(WebCore::AXIsolatedObject::colorAttributeValue const):
(WebCore::AXIsolatedObject::fontAttributeValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):

Canonical link: <a href="https://commits.webkit.org/294909@main">https://commits.webkit.org/294909@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/943307227f65f0897b6866a835acf044d7a63e7e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103335 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23011 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13332 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108508 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53978 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105374 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23351 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31562 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78528 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35480 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106341 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18104 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93220 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58861 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17940 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11261 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53334 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87742 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11323 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110889 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30473 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22499 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87529 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30838 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89417 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87166 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22218 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32026 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9744 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24807 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30402 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35721 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30208 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33535 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31770 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -800,9 +800,17 @@ void AXIsolatedTree::updateNodeProperties(AccessibilityObject& axObject, const A
-        case AXProperty::Font:
-            properties.append({ AXProperty::Font, axObject.font() });
+        case AXProperty::Font: {
+            if (RefPtr parent = axObject.parentObject()) {
```
Isn't this potentially a more expensive call? The code to compute parentObject() on the live tree is surprisingly complex sometimes.

hoffmanjoshua replied:
It's possible that we can get rid of this, now that we are checking the difference between self/parent _before_ calling `setProperty`. I'll investigate!

minorninth commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2039,6 +2039,16 @@ void AXObjectCache::onStyleChange(Element& element, OptionSet<Style::Change> cha
+        return;
+
+    if (!oldStyle->fontCascadeEqual(*newStyle))
+        tree->queueNodeUpdate(object->objectID(), { AXProperty::Font });
```
I missed this before, but I think that when there's a style change we have to recompute the style on the whole subtree, because we don't know whether descendants of this node might have had the same computed font and color before, but don't now.

hoffmanjoshua replied:
The good thing here is that `onStyleChange` will be called down the subtree (because inherited properties have changed), so the right properties should be updated here.

minorninth replied:
So here's the case I'm worried about:

```
<div id="a" style="color: black">
    <div id="b" style="color: black">
        What color is this text?
    </div>
</div>
```

We're only going to store "black" once on "a", and "b" will not store it because it gets it from its parent.

Let's say someone calls $("a").style.color = 'red'. The computed style of "b" doesn't actually change because it sets its own color.

Now if we query the accessible color of "b" we'll incorrectly get red (from its parent), when it hasn't actually changed.

hoffmanjoshua replied:
Oh this is a great point, yeah, we'll definitely need to do that (just threw this example into a file to verify).

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -1163,31 +1157,73 @@ Path AXIsolatedObject::pathAttributeValue(AXProperty property) const
-    );
+#ifndef NDEBUG
+    if (RefPtr parent = parentObject(); parent && property == AXProperty::TextColor)
+        RELEASE_ASSERT(parent->cachedTextColor() != getColor(m_properties[index].second));
```
Should this be a RELEASE_ASSERT?

hoffmanjoshua replied:
Fixed!

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -1163,31 +1157,73 @@ Path AXIsolatedObject::pathAttributeValue(AXProperty property) const
-    );
+#ifndef NDEBUG
+    if (RefPtr parent = parentObject(); parent && property == AXProperty::Font)
+        RELEASE_ASSERT(parent->cachedFont() != getFont(m_properties[index].second));
```
Should this be a RELEASE_ASSERT?

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -800,9 +800,17 @@ void AXIsolatedTree::updateNodeProperties(AccessibilityObject& axObject, const A
+            if (RefPtr parent = axObject.parentObject()) {
+                RetainPtr font = axObject.font();
+                if (font != parent->font()) {
+                    properties.append({ AXProperty::Font, font });
```
Can we WTFMove `font` here?

twilco commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -397,6 +397,10 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
     void onStyleChange(RenderText&, StyleDifference, const RenderStyle* oldStyle, const RenderStyle& newStyle);
+#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)
+    bool onFontChange(Element&, const RenderStyle*, const RenderStyle*);
+    bool onTextColorChange(Element&, const RenderStyle*, const RenderStyle*);
```
Can we add a comment to each of these explaining what the returned `bool` value means?

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2042,6 +2042,50 @@ void AXObjectCache::onStyleChange(Element& element, OptionSet<Style::Change> cha
+#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)
+bool AXObjectCache::onFontChange(Element& element, const RenderStyle* oldStyle, const RenderStyle* newStyle)
+{
+    if (!oldStyle || !newStyle)
+        return false;
+
+    RefPtr object = get(element);
+    if (!object)
+        return false;
```
Now, `onFontChange`, `onTextColorChange`, and `onStyleChange` repeat this work back-to-back. Can we avoid this?

One idea: move all of this back into `onStyleChange` and make it return a struct:

```cpp
struct AXStyleChangeResult {
    bool fontChanged;
    bool textColorChanged;
}
```

That solves all the problems — avoids duplicated work, and the undocumented bool value return issue (since we're now returning a struct with named fields, inherently self-documenting).

twilco commented on Source/WebCore/style/StyleTreeResolver.h:
```diff
@@ -116,6 +116,12 @@ class TreeResolver {
+        // Used to optimize accessibility isolated tree updates.
+        bool isAXFontChangeSubtree { false };
+        bool isAXColorChangeSubtree { false };
+#endif
```
This should go at the bottom of the structure definition, near all the other member variables.

Can you explain the name of these variables to me? I don't really understand, even after looking at how it's used. In any case, having `AX` as part of the name seems inaccurate. AX / accessibility is what consumes the bools, but not descriptive of the actual state it represents.

hoffmanjoshua replied:
These are part of the `Parent` struct, so they should to live here (under the other `Parent` member variables) and before the constructor. 

I'll update the name to be more reflective of the state and less specific to accessibility. They more generally mean "this subtree has a font/text change" and the names should reflect that.

hoffmanjoshua committed:
AX: AXIsolatedObject::setProperty does unsafe access to accessibility-thread only data structure, AXIsolatedTree::m_readerThreadNodeMap, when caching AXProperty::{Font, TextColor}
https://bugs.webkit.org/show_bug.cgi?id=292823
rdar://151077899

Reviewed by Tyler Wilcock.

294634@main made it possible for AXIsoaltedObject::parentObject() to be called on the main thread,
since it started making calls in setProperty in order to determine the parent object's style.

Instead of using the isoalted object, we can do the same diff-ing of Text Color and Font with live
objects. We need to do this in two places: when updating the property and during initialization.

To ensure we aren't updating the Font or Text Color property multiple times, we only update once
per subtree that had a Font/Color change, utilizing the parent stack in the style tree resolver.

At the same time, our new way of caching Font and Text Color using ancestors allows us to stop
caching these properties on static text, and instead cache them on any Element. This maintains
our previous optimization, and prevents possible duplication of properties.

This patch was co-authored by Tyler Wilcock.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::setProperty):
(WebCore::defaultTextColor):
(WebCore::getColor):
(WebCore::AXIsolatedObject::cachedTextColor const):
(WebCore::getFont):
(WebCore::AXIsolatedObject::cachedFont const):
(WebCore::AXIsolatedObject::colorAttributeValue const):
(WebCore::AXIsolatedObject::fontAttributeValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):

Canonical link: https://commits.webkit.org/294909@main
